src/sentry/runner/commands/backup.py

```diff
@@ -1,15 +1,105 @@
 from __future__ import annotations
 
+from difflib import unified_diff
 from io import StringIO
+from typing import NamedTuple, NewType
 
 import click
 from django.apps import apps
 from django.core import management, serializers
 from django.db import IntegrityError, connection, transaction
 
 from sentry.runner.decorators import configuration
+from sentry.utils.json import JSONData, JSONEncoder, better_default_encoder
 
 EXCLUDED_APPS = frozenset(("auth", "contenttypes"))
+JSON_PRETTY_PRINTER = JSONEncoder(
+    default=better_default_encoder, indent=2, ignore_nan=True, sort_keys=True
+)
+
+ComparatorName = NewType("ComparatorName", str)
+ModelName = NewType("ModelName", str)
+
+
+# TODO(team-ospo/#155): Figure out if we are going to use `pk` as part of the identifier, or some other kind of sequence number internal to the JSON export instead.
+class InstanceID(NamedTuple):
+    """Every entry in the generated backup JSON file should have a unique model+pk combination, which serves as its identifier."""
+
+    model: ModelName
+    pk: int
+
+    def pretty(self) -> str:
+        return f"InstanceID(model: {self.model!r}, pk: {self.pk})"
+
+
+class ComparatorFinding(NamedTuple):
+    """Store all information about a single failed matching between expected and actual output."""
+
+    name: ComparatorName
+    on: InstanceID
+    reason: str = ""
+
+    def pretty(self) -> str:
+        return f"Finding(\n\tname: {self.name!r},\n\ton: {self.on.pretty()},\n\treason: {self.reason}\n)"
+
+
+class ComparatorFindings:
+    """A wrapper type for a list of 'ComparatorFinding' which enables pretty-printing in asserts."""
+
+    def __init__(self, findings: list[ComparatorFinding]):
+        self.findings = findings
+
+    def append(self, finding: ComparatorFinding) -> None:
+        self.findings.append(finding)
+
+    def pretty(self) -> str:
+        return "\n".join(f.pretty() for f in self.findings)
+
+
+def validate(expect: JSONData, actual: JSONData) -> ComparatorFindings:
```

Comment:
wondering if we want this to be a cli command that users can run after importing

Reply:
I think eventually I'd like to get to a place where this is an automatic part of the import process - the import is not "committed" if it does not validate. Allowing users to run validation independently feels pretty error prone to me - if it fails after you import, you already have a corrupted database.